### PR TITLE
Added functionality to launch PyGrid network and nodes upon creation of a cluster

### LINKED ISSUE #3099: Implement Auto-Scaling on Google Cloud
### Description

In this project, you will implement functionality necessary to automatically spin-up Google cloud machines, load a PyGrid instance, run a training job, and tear down the instance upon completion (depositing the results into another long-running instance). The primary feature will be the ability to run a “hyperparameter sweep”, as exemplified below.

```python
parameters = {“alpha” : [0, 0.01, 0.02], “batch_size” : [32,64]}

gcloud = gr.GoogleCloud(credentials)

cluster = gcloud.LazyCluster(n_machines=10, 
                            type="n-series", 
                            priority="low",
                            eviction_policy="Delete",
                            max_price_usd=0.55,
                            reboot_max_price_usd=0.65)

cluster.sweep(model=model,
             parameters=parameters,
             optim=optim.SGD,
             results_node=grid['my node']
             )
```

### Context

I will be mentoring this project as part of the [Google Summer of Code](https://summerofcode.withgoogle.com/)

Also, this year we have some plans for PyGrid, along with our other projects, including a focus on production: Deployment on Cloud Providers (GCP, Amazon, Azure). You can check out the PyGrid Roadmap [here](https://github.com/OpenMined/Roadmap/tree/master/pygrid_team).

### Required Skills:

- Knowledge of [PyTorch](https://pytorch.org/) and Deep Learning
- Familiarity (or willingness to become familiar) with [PySyft](https://github.com/OpenMined/PySyft)
- Familiarity (or willingness to become familiar) with [Google Cloud’s Infrastructure](https://cloud.google.com/docs)

**Difficulty**: Beginner

While this is a sizeable project, the core functionality is relatively straightforward (provision machines, install and start PyGrid servers, and run training scripts). Furthermore, there are plenty of tutorials on how to run PyGrid servers and how to run Google Cloud machines.

If you are new to PyGrid and want to learn more about it, try out the tutorials. A comprehensive list of tutorials can be found [here](https://github.com/OpenMined/PySyft/tree/master/examples/tutorials/grid). These tutorials cover how to create a PyGrid node and what operations you can perform.

### Useful Links

- PyTorch - https://pytorch.org/blog
- PySyft - https://github.com/OpenMined/PySyft
- PyGrid - https://github.com/OpenMined/PyGrid
- GCP Docs - https://cloud.google.com/docs
- GCloud APIs for Python - https://cloud.google.com/python/docs/reference
- Google APIs (Python Client) - https://github.com/googleapis/google-api-python-client 
- Google Summer of Code - https://summerofcode.withgoogle.com
  - [Timeline](https://developers.google.com/open-source/gsoc/timeline)
  - [Student Guide](https://google.github.io/gsocguides/student/)
  - [OpenMined Org.](https://summerofcode.withgoogle.com/organizations/5561599089180672)

_List of all GSoC project ideas [here](https://docs.google.com/document/d/1OA8277RB9ZN8u7Wgn6NqZK7tzixnZbSlkfulyqoo2G4/edit?usp=sharing)._

### PR DESCRIPTION
**Pull request by rimijoker**
## Description
I have added startup-scripts and other necessary changes which will enable the autoscale API to automatically launch PyGrid servers(network and nodes). Now using the autoscale API can launch PyGird cluster(1 network + N nodes) with one function call or create a PyGrid network and add PyGrid nodes to it one at a time as per the requirement.

I have also added tests and an example tutorial training script for the demonstration of the same.

@hericlesme 
Issue #3099

## Affected Dependencies
None (New Feature)

## How has this been tested?
- I have created test.py script and test.ipynb notebook to test the functionality of the features implemented 
- I have also modified the README.md with examples on how to use the functionality. 
- I have also modified a PyGrid tutorial to use this API for demonstration.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labelled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: syft/grid/utils/autoscale/README.md [hunk 1]
```diff
@@ -71,14 +77,35 @@ instance_name.compute_instance(
...
+
+```python
+instance_name.create_gridnetwork(
+    name="new-network",
+    machine_type=configs.MachineType.f1_micro,
+    zone=configs.Zone.us_central1_a,
+)
+```
+
+- Create PyGrid Node instance using :
```
Comment (hericlesme) [inline syft/grid/utils/autoscale/README.md:90]: Remove spacing between phrase and colon

File: syft/grid/utils/autoscale/README.md [hunk 2]
```diff
@@ -71,14 +77,35 @@ instance_name.compute_instance(
...
+```python
+instance_name.create_gridnode(
+    name="new-node",
+    machine_type=configs.MachineType.f1_micro,
+    zone=configs.Zone.us_central1_a,
+    gridnetwork_name="new-network",
+)
+```
+
 - Create Clusters using :
```
Comment (hericlesme) [inline syft/grid/utils/autoscale/README.md:101]: Remove spacing between phrase and colon

File: syft/grid/utils/autoscale/gcloud.py [hunk 3]
```diff
@@ -34,23 +35,145 @@ def __init__(self, credentials, project_id, region):
...
+    def expose_port(self, name, ports=None, apply=True):
+        """
+        args:
+            name: name of the resource
+            ports: list of ports to be exposed, defaults to 80
+            apply: to call terraform apply at the end
+        """
+        if not ports:
+            ports = [80]
+
```
Comment (hericlesme) [inline syft/grid/utils/autoscale/gcloud.py:47]: Why not use port as a default parameter value? Like

```python
def expose_port(self, name, ports=[80], apply=True):
    ...
```
Also, if you want to keep the current way we could do:
```python
if ports is not None:
    ports = [80]
```

File: syft/grid/utils/autoscale/gcloud.py [hunk 4]
```diff
@@ -34,23 +35,145 @@ def __init__(self, credentials, project_id, region):
...
+            zone=zone,
+            boot_disk={"initialize_params": {"image": "${" + image.self_link + "}"}},
             network_interface={"network": "default", "access_config": {}},
+            metadata_startup_script=f"""
+                #!/bin/bash
+                apt-get update
+                apt-get -y upgrade
+                sudo -i bash -c 'pip install notebook==5.7.8'
+                sudo -i bash -c 'pip install git+https://github.com/OpenMined/PyGridNode.git'
+                sudo -i bash -c 'echo Starting Node {name} \
```
Comment (Nilanshrajput) [inline syft/grid/utils/autoscale/gcloud.py:172]: GridNode Repo is archived now, this should be updated to new method for creating nodes from Pygrid

Comment (rimijoker) [reply syft/grid/utils/autoscale/gcloud.py:172]: Now as the refactoring has completed, I will update it to the new methods shortly.
